01-architecture.md

```diff
@@ -0,0 +1,222 @@
+
+Approach 2 is simpler, and places no constraints on the sandbox. It is easier to reason about and reduces overall complexity. It's also more performant because the sandbox has no need to analyse the WASM exports table. Evolving actors over time (such as introducing interface changes) is also easier to reason about.
+
+Finally, this approach is readily compatible with VMs that rely on internal dispatch (e.g. EVM), but also stretches to accommodates for VMs that perform outer dispatch.
+
+Approach 2 would render the `MethodNum` field on messages obsolete, as external dispatch is no longer necessary, and all call information would be contained in the input parameters.
+
+We _could_ technically preserve the `MethodNum`, and have actors perform internal dispatch based on that _and_ input parameters. But we believe that's unnecessary cruft, and it breaks the actor-orientation paradigm by carrying over a procedural construct.
+
+Not everything is rosy, though. The main risk with Approach 2 is that the calling convention not explicited, which may lead to proliferation of calling conventions across user-deployed actors. In turn this may result in cognitive overhead and interoperability issues.
+
+An IPLD-based Interface Definition Language should eliminate that concern. 
+
+**Future: IPLD-based IDLs**
+
+IPLD Schemas are good for defining the structure of objects, but they do not help with expressing the interface of behaviours an object can offer (e.g. methods), inputs, and outputs.
+
+Our basic idea is to design an IPLD schema for IDLs; think of [gRPC IDLs](https://grpc.io/docs/what-is-grpc/core-concepts/), [SOAP WSDL](https://www.w3schools.com/xml/xml_wsdl.asp), etc. This schema would be standardised and versioned, and understood by the FVM SDK.
+
+This IDL would not have functions as such, but rather _labelled behaviours_. This is consistent with the actor model, where actors don't expose functions, but they are known to provide behaviours.
+
+Each behaviour would carry an IPLD Schema for the input type and the return value type, an error code enum, and potentially an identifier (which incidentally could be a method number!) to facilitate internal dispatch.
+
+Traditional "overloaded functions" can be represented as distinct labelled behaviours dispatching to the same identifier.
+
+The FVM SDK would offer utilities to pack/generate calls to another actor with their IDL.
+
+_Pseudocode:_
+
+```
+import "idls/other-actor.idl.ipld" as actor
+from sdk import calls
+
+// Res is typed automatically as Result<Type, Error>
+let res = calls.target(actor)
+               .call("behaviour")
+               .with({ Field2: Value, Field2: [Value1, Value2]})
+```
+
+**Test sandbox**
+
+The FVM project will need to ship a test actor sandbox for developers to use during their actor development process.
+
+**Foreign runtimes**
```

Comment:
Are we just talking other _blockchain_ runtimes, or runtimes in general? (EVM or JVM/PVM/node/etc.)

Reply:
This was really in the context of my questions below (i.e., blockchain-based runtimes are going to be harder because they'll interact with the chain/other actors while JVM etc. doesn't need that kind of built-in support).